src/ch/ch_monitor.c

```diff
@@ -563,25 +564,341 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+static inline char *end_of_json(char *str, size_t len)
+{
+    bool started = false;
+    int blocks = 0;
+    int i = 0;
+
+    while ((i < len) && (!started || blocks > 0)) {
+        if (str[i] == '{') {
+            if (!started)
+                started = true;
+            blocks++;
+        } else if (str[i] == '}') {
+            blocks--;
+        }
+        i++;
+    }
+
+    return (i == len && blocks) ? NULL : str + i;
+}
+
+/*
+ * Caller should have reference on Monitor and Domain
+ */
+static int virCHMonitorProcessEvents(virCHMonitorPtr mon, int events)
+{
+    virJSONValuePtr obj = NULL;
+    char *buf = mon->buffer;
+    int ret = 0;
+    int i = 0;
+    size_t sz;
+
+    for (i = 0; i < events; i++) {
+        char tmp;
+        char *end = end_of_json(buf, mon->buf_fill_sz);
+
+        /*
+         * end should never be NULL! We validated that there
+         * is a valid JSON document before calling end_of_json,
+         * and end_of_json returns NULL only if it cannot find
+         * a valid JSON document.
+         */
+        assert(end);
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wnull-dereference"
+        tmp = *end, *end = 0;
+#pragma GCC diagnostic pop
+
+        if ((obj = virJSONValueFromString(buf))) {
+            if (virCHMonitorProcessEvent(mon, obj) < 0) {
+                VIR_WARN("Failed to process the event!");
+                ret = -1;
+            }
+            virJSONValueFree(obj);
+        } else {
+            VIR_WARN("Invalid JSON from monitor");
+            ret = -1;
+        }
+
+        *end = tmp;
+        buf = end;
+    }
+
+    /*
+     * If the buffer still has incomplete data, lets
+     * push it to the beginning.
+     */
+    sz = buf - mon->buffer;
+    if (sz < mon->buf_fill_sz) {
+        mon->buf_offset = mon->buf_fill_sz - sz;
+        memmove(mon->buffer, buf, sz);
+    } else {
+        mon->buf_offset = 0;
+    }
+
+    return ret;
+}
+
+static int virCHMonitorReadProcessEvents(virCHMonitorPtr mon)
+{
+    size_t max_sz = CH_MONITOR_BUFFER_SZ - mon->buf_offset;
+    char *buf = mon->buffer + mon->buf_offset;
+    virDomainObjPtr vm = mon->vm;
+    bool incomplete = false;
+    int events = 0;
+    size_t sz = 0;
+
+    memset(buf, 0, max_sz);
+    do {
+        ssize_t ret;
+
+        ret = read(mon->monitor_fd, buf + sz, max_sz - sz);
+        if (ret == 0 || (ret < 0 && errno == EINTR)) {
+            g_usleep(G_USEC_PER_SEC);
+            continue;
+        } else if (ret < 0) {
+            /*
+             * We should never reach here. read(2) says possible errors
+             * are EINTR, EAGAIN, EBADF, EFAULT, EINVAL, EIO, EISDIR
+             * We handle EINTR gracefully. There is some serious issue
+             * if we encounter any of the other errors(either in our code
+             * or in the system). Better to bail out.
+             */
+            VIR_ERROR("Failed to read monitor events!: %s", strerror(errno));
+            abort();
```

Comment:
Thinking about this more, I'm not sure if this should abort since this could happen if a cloud-hypervisor process is killed for instance and the abort would cause the entire libvirt process to go down with a single vm. Perhaps just stopping the event loop, changing the vm's status to shutdown and returning would be correct behavior for this case?

Reply:
I think failure of read() system call on this pipe should be considered fatal. cloud-hypervisor process dying will not return error on read I think. It would be like encountering eof. read can return EAGAIN, EISDIR, EFAULT, EBADF and all these errors would be caused only due to a programming error in our driver. It can return EIO which could be caused wither by a programming error or a deep kernel level problem. Other than that EINTR can happen and we take care of that.

I am approaching it as a failfast mechanism where we detect the bug in our code fast by aborting libvirtd and then fix the problem in code fast. I agree EIO is out of our hands and probably should handled like what you mentioned. But for rest of the cases I think its better to crash libvirtd so that we as developers get to know the problem if it happens.

Also killing of the libvirtd should not kill the VMs. Since we run each VM as a daemon and parented to init, VMs should run unaffected. Restarting libvirtd would bring everything back to normal I guess.

What do you feel?

Reply:
Ah yes you are right, it would hit an EOF since it can read from a closed pipe up to however much data was left over.

For the EOF case, I'm not sure if we shouldn't just stop the loop as we just are adding wake-ups to the system every second otherwise but I don't think it makes much difference so I'm fine with the code as is.

Reply:
EOF case is already taken care off I guess(if virDomainObjIsActive(vm) returns false when cloud-hypervisor gets killed). As of now, Normall kill command (SIGTERM) works as expected. But SIGKILL doesn't notify libvirt. We would need to fix this and then the loop will be exited with out any extra effort. 
We have an action item here to make libvirt aware of cloud-hypervisor abnormal exits. Filed and issue here: https://github.com/cloud-hypervisor/libvirt/issues/47

Reply:
RIght, of course. Thank you so much for going through all this. Quite a lot of details to get right X(.